src/modules/iotjs_module_dbus.c

```diff
@@ -423,22 +422,48 @@ JS_FUNCTION(GetBus) {
   IOTJS_VALIDATED_STRUCT_METHOD(iotjs_dbus_t, dbus);
 
   DBusError error;
-  dbus_error_init(&error);
-
   DJS_CHECK_ARGS(1, number);
   int type = JS_GET_ARG(0, number);
-  if (type == 0 /* BUS_SYSTEM */) {
-    _this->connection = dbus_bus_get(DBUS_BUS_SYSTEM, &error);
-  } else if (type == 1 /* BUS_SESSION */) {
-    _this->connection = dbus_bus_get(DBUS_BUS_SESSION, &error);
+  int wait = 1000; // default is 1s
+  if (jerry_value_is_number(jargv[1])) {
+    wait = iotjs_jval_as_number(jargv[1]);
   }
+  wait *= 1000;
+
+  // FIXME(Yorkie): currently the implementation waits synchronously because it
+  // might be used in edge-case, when the dbus-daemon is unavailable or the unix
+  // path is not found.
+  do {
+    dbus_error_init(&error);
+    if (type == 0 /* BUS_SYSTEM */) {
+      _this->connection = dbus_bus_get(DBUS_BUS_SYSTEM, &error);
+    } else if (type == 1 /* BUS_SESSION */) {
+      _this->connection = dbus_bus_get(DBUS_BUS_SESSION, &error);
+    }
+
+    if (_this->connection != NULL)
+      break;
+
+    fprintf(stderr, "DBUS(%s) %s\n", error.name, error.message);
+    if (wait <= 0 ||
+        (strcmp(error.name, "org.freedesktop.DBus.Error.FileNotFound") != 0 &&
+         strcmp(error.name, "org.freedesktop.DBus.Error.Failed") != 0)) {
+      break;
+    } else {
+      usleep(IOTJS_DBUS_CONN_DELAY);
```

Comment:
Is this `usleep` sleeping on main thread?

Reply:
See https://github.com/Rokid/ShadowNode/pull/408/files#diff-ddc1ef60d9b05e87f82ea405bdf4e9aaR433, yes, it blocks the main thread the `max_timeout` ms in bad case.

Reply:
If it is more suitable that throwing an error to JS and use setTimeout which is based on uv timers? or directly register an uv timer in C land.

Reply:
agreed with @legendecas

Reply:
Breaking API is more acceptable to me to sleep in a library. Is there any other strong objections against throwing an error or provide an async API?

Reply:
> Breaking API is more acceptable to me than sleeping in a library

We should discuss on a context, non-blocking API might be better than current implementation is that I was agreeing, too. That's why I commented a FIXME on my source code. The context is that introducing a breaking change causes that could not be landed at v0.10.x.

Reply:
If a new async api shall not be introduced in a patch version, an error shall be thrown rather than blocking implicitly which would definitely introduce unpredictable behaviors.

Reply:
Error handling in establishing a connection might not be an edge case as the comment saying IMO.